Climdex dialog updated for multiple stations

dannyparsons opened the pull request:
- added wrapper climdex R functions for multiple stations
- updated dialog to new functions

@africanmathsinitiative/developers this is ready for review

**Please test:**
- Annual and Monthly options
- With and without multiple stations
- That it works with different kinds of year/month column e.g. numeric or factor
- Running multiple times to see if summaries go in the same sheet (if the same station/year/month columns are used)
- Doing other summaries at the same level before and/or after using the dialog to see if they go in the sheet e.g. Climatic Summaries followed by Climdex

lloyddewit commented on instat/static/InstatObject/R/instat_object_R6.R:
```diff
@@ -1349,111 +1349,132 @@ DataBook$set("public", "remove_key", function(data_name, key_name) {
+        if (length(mns) == 12) {
+          if (setequal(mns, month.abb)) climdex_output[[month]] <- month.abb[climdex_output[[month]]]
+          else if (setequal(mns, month.name)) climdex_output[[month]] <- month.name[climdex_output[[month]]]
+          else if (setequal(mns, month.name.english)) climdex_output[[month]] <- month.abb.english[climdex_output[[month]]]
```
The if condition is the same as the line below. Should we check against `month.abb.english` instead?

dannyparsons replied:
Yes, thanks.

lloyddewit commented on instat/static/InstatObject/R/instat_object_R6.R:
```diff
@@ -1349,111 +1349,132 @@ DataBook$set("public", "remove_key", function(data_name, key_name) {
+      else if (is.character(col_month)) {
+        mns <- unique(col_month)
+        # Also check English names as month.abb and month.name are locale dependent.
+        month.abb.english <- c("Jan","Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
+        month.name.english <- c("January", "February", "March", "April", "May", "June", "July", 
+                                "August", "September", "October", "November", "December")
```
Do we need to consider case sensititivity?

dannyparsons replied:
This is how they would be created from R automatically, but they could be in other forms. I've added checks for all upper case or all lower case as well.

lloyddewit commented on instat/static/InstatObject/R/instat_object_R6.R:
```diff
@@ -1349,111 +1349,132 @@ DataBook$set("public", "remove_key", function(data_name, key_name) {
+        if (length(mns) == 12) {
+          if (setequal(mns, month.abb)) climdex_output[[month]] <- month.abb[climdex_output[[month]]]
+          else if (setequal(mns, month.name)) climdex_output[[month]] <- month.name[climdex_output[[month]]]
+          else if (setequal(mns, month.name.english)) climdex_output[[month]] <- month.abb.english[climdex_output[[month]]]
```
The if condition is the same as the line below. Should we check against `month.abb.english` instead?

dannyparsons replied:
Yes thanks.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+                   "fd" = climdex.pcic::climdex.fd(ci),
+                   "su" = climdex.pcic::climdex.su(ci),
+                   "id" = climdex.pcic::climdex.id(ci),
+                   "tr" = climdex.pcic::climdex.id(ci),
```
is this correct?

dannyparsons replied:
Fixed thanks.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+                   "tx90p" = climdex.pcic::climdex.tx90p(ci, freq = freq),
+                   "dtr" = climdex.pcic::climdex.dtr(ci, freq = freq),
+                   "rx1day" = climdex.pcic::climdex.rx1day(ci, freq = freq),
+                   "rx5day" = climdex.pcic::climdex.rx1day(ci, freq = freq),
```
is this correct?

dannyparsons replied:
Fixed thanks.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+      if (freq == "annual") {
+        df_ind <- data.frame(names(vals), unname(vals))
+        names(df_ind) <- c(year, indices[i])
+      } else if(freq == "monthly") {
```
```suggestion
      } else {
```
If we do check for `"monthly"` then for consistency should we also add a `else {stop ...`?

dannyparsons replied:
Agreed, there is a check and stop above to make sure it is one of the two values so `else` is fine.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+  stopifnot(freq %in% c("annual", "monthly"))
+  if (freq == "monthly" && missing(month)) stop("month is required for freq = 'monthly'.")
+  
+  # All indices can only be calculated annually. Only some have monthly versions as well.
```
```suggestion
  # All indices can be calculated annually. Only some have monthly versions as well.
```

dannyparsons replied:
Done.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+  if (freq == "monthly" && missing(month)) stop("month is required for freq = 'monthly'.")
+  
+  # All indices can only be calculated annually. Only some have monthly versions as well.
+  year_only_indices <- c("fd", "su", "id", "tr", "wsdi", "csdi", "gsl", "sdii", "r10mm", 
```
These literal strings are duplicated. The compiler won't notice typos. Use constants?

dannyparsons replied:
This is what we should do when we make an R package with these functions. It's not so standard to make constants otherwise.

lloyddewit replied:
OK, understood

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+    else if (is.character(data[[station]])) df_out[[station]] <- as.character(df_out[[station]])
+    else warning("Cannot recognise the class of station column. Link between data frames may be unstable.")
+  }
+  if (is.numeric(data[[year]])) df_out[[year]] <- as.numeric(df_out[[year]])
+  else if (is.factor(data[[year]])) df_out[[year]] <- make_factor(df_out[[year]])
+  else if (is.character(data[[year]])) df_out[[year]] <- as.character(df_out[[year]])
```
This type of logic seems to reoccur in the previous R file. Place this logic in a shared function?

dannyparsons replied:
Yes, you're right. I've put it as a TODO comment for now.

lloyddewit commented on instat/static/InstatObject/R/stand_alone_functions.R:
```diff
@@ -1565,3 +1565,149 @@ make_factor <- function(x, ordered = is.ordered(x)) {
+    else if (is.character(data[[month]])) {
+      mns <- unique(data[[month]])
+      # Also check English names as month.abb and month.name are locale dependent.
+      month.abb.english <- c("Jan","Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")
+      month.name.english <- c("January", "February", "March", "April", "May", "June", "July", 
+                              "August", "September", "October", "November", "December")
```
duplicated literals, make constants?

dannyparsons replied:
Done.

lloyddewit reviewed (changes requested):
Just some small comments on the R code

dannyparsons commented:
@lloyddewit thanks for picking up those detailed mistakes. I've addressed your comments, could you check the latest changes?

@rdstern could review functionality in parallel.

rdstern reviewed (changes requested):
@dannyparsons this is great.  I have been trying it on the Niger (awkward) data and so far just about so good.
1) On the settings tab it used to say 1961 to 1990 as the base.  Now it said 1900 to 1900.  Perhaps that was to do with my data?
2) More serious is the settings for the maximum number of missing values.  The minimums of the up-down are both set to 1 and I suggest should be zero?  Second I don't think they are changing when I changed the settings down as far as I could go.  I failed to check the command though.

dannyparsons commented:
Thanks @rdstern those were both bugs, now fixed.

lloyddewit reviewed (commented):
@dannyparsons FYI, I reviewed but have no comments
